### PR TITLE
fix(dot/network): fix memory allocations with `sizedBufferPool`

### PR DESCRIPTION
**Pull request by qdm12**
## Changes

- Use a slice instead of an array, since returning the array was effectively copying it

## Tests

```
go test -benchmem -run=Benchmark_sizedBufferPool github.com/ChainSafe/gossamer/dot/network
```

## Issues

- Related to #1936 and found by reviewing PR #1913 

## Primary Reviewer

- Anyone is welcome

### CODE REVIEW

**Comment by danforbes** [comment]
Is there any way to quantify the impact/improvement this has the performance of the code?

**Comment by qdm12** [comment]
> Is there any way to quantify the impact/improvement this has on the performance of the code?

Not really, although it is now significantly better, see [benchmark results](https://github.com/ChainSafe/gossamer/pull/1963/files#diff-1c450d8732ec5666cc3fa5ae5d2ef3b2b15b3957c9de12d2037bd50d78c45bccR18-R21). Since that piece was used for all dot networking, it could either had a very bad impact or it would had been barely noticeable. Only way to find out is to run it 😄 (I mean it, and check memory usage over time, profiling won't help really here I think)

**Comment by danforbes** [comment]
Benchmark results are super quantifiable...thanks, @qdm12 🙏🏻

**Comment by qdm12** [comment]
@danforbes but let's not get too excited. Go's runtime is pretty efficient even with bad code, so there might not be any concrete difference sadly. Anyway, one can only hope 🌟 !

**Comment by noot** [comment]
@qdm12 how does the CPU usage get affected by this (if any)? as the main point of the buffer pool was to prevent unnecessary malloc's which used a lot of CPU

**Comment by qdm12** [comment]
@noot the previous implementation makes sense but was bugged. It would copy the array on every call to `get()` so effectively it was useless (the same as declaring a new array on every call). See the benchmark results commented below the benchmark.

I benchmarked using an array pointer and a slice (which has a pointer to the underlying data), and both gave similar results, so I decided to go with the slice for simplicity.

Side question, did we notice a performance bottleneck in this part? We should avoid such performance improvements without identifying a bottleneck and making sure there is a visible difference of performance (don't get me wrong, I love optimising 😄). That pool being bugged, it must had been slower than the original code, and we still used it; my point is we should be more careful about such optimisation changes in the future 👍

**Comment by noot** [comment]
@qdm12 sorry, did you bench the CPU or just memory? yes, this was created because previously we were allocating a new buffer every time we received a network message, which was causing gossamer to use 100% CPU on every core basically immediately after starting. the CPU usage improved dramatically after adding this

**Comment by qdm12** [comment]
> did you bench the CPU or just memory

It does both.

> I guess more ops in the same amount of time

Yes. You can see before it was using a lot more time (and memory) per operation (op).

> the CPU usage improved dramatically after adding this ... which was causing gossamer to use 100% CPU on every core basically immediately after starting

That doesn't make sense, since the array was copied to another memory location on every `get` request. It might had been something else resolving the 100% cpu performance issue. I also doubt the GC (or malloc) would use 100% of even a single CPU core, even under high load. Also the network is magnitudes slower than memory bandwidth.
For science's sake 🥼, and once we have something easy to deploy changes, I would be curious to try reverting it and check CPU usage 😄

**Comment by noot** [comment]
@qdm12 I'm just going based off the profiling lol, it showed malloc using up all the CPU it could previously. I can try to find an old commit for you to try to profile if you'd like ;)

**Review by noot — CHANGES_REQUESTED** [review]
[redacted] nevermind, let me test this more

File: dot/network/service.go [hunk 1]
```diff
@@ -149,13 +149,11 @@ func NewService(cfg *Config) (*Service, error) {
...
-		bufPool = &sizedBufferPool{
-			c: make(chan *[maxMessageSize]byte, cfg.MinPeers*3),
-		}
-	} else {
-		bufPool = newSizedBufferPool(cfg.MinPeers*3, cfg.MaxPeers*3)
+	preAllocatedInPool := 0
+	if !cfg.noPreAllocate {
+		preAllocatedInPool = cfg.MinPeers * 3
 	}
+	bufPool = newSizedBufferPool(preAllocatedInPool, cfg.MinPeers*3)
```
Comment (noot) [inline dot/network/service.go:156]: the maximum should be `cfg.MaxPeers*3`

Comment (qdm12) [reply dot/network/service.go:156]: Ohhh thank you, blind mee!! I'll fix that.

Comment (qdm12) [reply dot/network/service.go:156]: Wait that's confusing too, why would the channel capacity `make(chan *[maxMessageSize]byte, cfg.MinPeers*3),` above not be `cfg.MaxPeers*3` as well then? Pre-allocation or not, I don't see how this should influence the channel capacity? 🤔

Comment (noot) [reply dot/network/service.go:156]: you mean in the `!cfg.noPreAllocate` case? that's only used for testing, so there isn't really a point in pre-allocating `cfg.MaxPeers*3` buffers

Comment (qdm12) [reply dot/network/service.go:156]: No, I mean why is the maximum channel capacity (aka `max`, different from the pre-allocated buffers in the channel queue) different.

When noPreAllocated (production) is true, the capacity is `cfg.MinPeers*3`. When it's false (for testing), it's `cfg.MaxPeers*3` 🤔  That looks strange right? Didn't we want to pre-allocate `cfg.MinPeers*3` and have a max of `cfg.MinPeers*3` for testing as well?

Comment (noot) [reply dot/network/service.go:156]: `noPreAllocated` is always false in production, as we want to pre-allocate, it's only set to true in the tests. in testing, it's fine to have a maximum channel capacity of `cfg.MinPeers*3` as we won't really be using many buffers (we could probably set it to 3 or something even). when it's false, in production, we want to be able to store `cfg.MaxPeers*3` buffers potentially as that is the maximum amount of buffers we woudl ever need (# of peers * 3 streams per peer)

Comment (noot) [reply dot/network/service.go:156]: @qdm12 can you update this when you get a chance? everything else looks good :)

Comment (qdm12) [reply dot/network/service.go]: Done :+1: Yeah I had it all wrong, it should be fixed now.

**Review by arijitAD — APPROVED** [review]

**Review by noot — APPROVED** [review]

**Pull request merged**